File: backend/app/ai/deepseek.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Optional

import requests


@dataclass
class DeepSeekConfig:
    base_url: str
    api_key: str
    model: str
    embed_base_url: str
    embed_model: str


class DeepSeekClient:
    def __init__(self, config: DeepSeekConfig) -> None:
        self._config = config
        self._session = requests.Session()
        self._timeout = 20

    def chat(self, system_prompt: str, user_prompt: str) -> str:
        endpoint = f"{self._config.base_url.rstrip('/')}/chat/completions"
        payload = {
            "model": self._config.model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            "max_tokens": 500,
            "temperature": 0.4,
            "stream": False,
        }
        headers = {
            "Authorization": f"Bearer {self._config.api_key}",
            "Content-Type": "application/json",
        }

        last_error: Optional[Exception] = None
        for attempt in range(1, 4):
            try:
                response = self._session.post(
                    endpoint, json=payload, headers=headers, timeout=self._timeout
                )
                if response.status_code >= 400:
                    try:
                        body = response.json()
                        message = body.get("error", {}).get("message")
                    except Exception:
                        message = None
                    if message:
                        raise RuntimeError(f"deepseek error: {message}")
                    raise RuntimeError(f"deepseek error: status {response.status_code}")

                data = response.json()
                choices = data.get("choices") or []
                if not choices:
                    raise RuntimeError("deepseek returned no choices")
                content = choices[0].get("message", {}).get("content")
                if not content:
                    raise RuntimeError("deepseek returned empty content")
                return str(content)
            except Exception as exc:  # pragma: no cover - best effort retries
                last_error = exc
                if attempt < 3:
                    time.sleep(attempt)
                else:
                    break
        raise RuntimeError(str(last_error) if last_error else "deepseek request failed")

    def embed(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        endpoint = f"{self._config.embed_base_url.rstrip('/')}/embeddings"
        payload = {
            "model": self._config.embed_model,
            "input": texts,
        }
        headers = {
            "Authorization": f"Bearer {self._config.api_key}",
            "Content-Type": "application/json",
        }

        last_error: Optional[Exception] = None
        for attempt in range(1, 4):
            try:
                response = self._session.post(
                    endpoint, json=payload, headers=headers, timeout=self._timeout
                )
                if response.status_code >= 400:
                    try:
                        body = response.json()
                        message = body.get("error", {}).get("message")
                    except Exception:
                        message = None
                    if message:
                        raise RuntimeError(f"deepseek error: {message}")
                    raise RuntimeError(f"deepseek error: status {response.status_code}")

                data = response.json()
                items = data.get("data") or []
                if not items:
                    raise RuntimeError("deepseek returned no embeddings")
                embeddings_by_index: dict[int, list[float]] = {}
                for item in items:
                    embedding = item.get("embedding")
                    if embedding is None:
                        continue
                    index = int(item.get("index", len(embeddings_by_index)))
                    embeddings_by_index[index] = embedding
                if not embeddings_by_index:
                    raise RuntimeError("deepseek returned empty embeddings")
                return [embeddings_by_index[i] for i in sorted(embeddings_by_index.keys())]
            except Exception as exc:  # pragma: no cover - best effort retries
                last_error = exc
                if attempt < 3:
                    time.sleep(attempt)
                else:
                    break
        raise RuntimeError(str(last_error) if last_error else "deepseek request failed")

File: backend_go/internal/ai/deepseek.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"
)

type AIChatBot interface {
	Chat(ctx context.Context, systemPrompt, userPrompt string) (string, error)
}

type Logger interface {
	Printf(format string, args ...any)
}

type DeepSeekClient struct {
	baseURL    string
	apiKey     string
	model      string
	httpClient *http.Client
	logger     Logger
}

type deepSeekMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type deepSeekRequest struct {
	Model       string            `json:"model"`
	Messages    []deepSeekMessage `json:"messages"`
	MaxTokens   int               `json:"max_tokens,omitempty"`
	Temperature float64           `json:"temperature,omitempty"`
	Stream      bool              `json:"stream"`
}

type deepSeekResponse struct {
	Choices []struct {
		Message deepSeekMessage `json:"message"`
	} `json:"choices"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error,omitempty"`
}

func NewDeepSeekClient(baseURL, apiKey, model string) *DeepSeekClient {
	trimmed := strings.TrimRight(baseURL, "/")
	return &DeepSeekClient{
		baseURL: trimmed,
		apiKey:  apiKey,
		model:   model,
		httpClient: &http.Client{
			Timeout: 20 * time.Second,
		},
		logger: log.New(io.Discard, "", 0),
	}
}

func (c *DeepSeekClient) Chat(ctx context.Context, systemPrompt, userPrompt string) (string, error) {
	if c == nil {
		return "", errors.New("deepseek client not configured")
	}
	const maxAttempts = 3
	var lastErr error

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		if err := ctx.Err(); err != nil {
			return "", err
		}

		endpoint := c.baseURL + "/chat/completions"
		payload := deepSeekRequest{
			Model: c.model,
			Messages: []deepSeekMessage{
				{Role: "system", Content: systemPrompt},
				{Role: "user", Content: userPrompt},
			},
			MaxTokens:   500,
			Temperature: 0.4,
			Stream:      false,
		}
		body, err := json.Marshal(payload)
		if err != nil {
			return "", err
		}

		c.logger.Printf("deepseek request attempt=%d system=%q user=%q", attempt, systemPrompt, userPrompt)
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
		if err != nil {
			return "", err
		}
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
		req.Header.Set("Content-Type", "application/json")

		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = err
		} else {
			var decoded deepSeekResponse
			decodeErr := json.NewDecoder(resp.Body).Decode(&decoded)
			_ = resp.Body.Close()
			if decodeErr != nil {
				lastErr = decodeErr
			} else if resp.StatusCode >= 400 {
				if decoded.Error != nil && decoded.Error.Message != "" {
					lastErr = fmt.Errorf("deepseek error: %s", decoded.Error.Message)
				} else {
					lastErr = fmt.Errorf("deepseek error: status %d", resp.StatusCode)
				}
			} else if len(decoded.Choices) == 0 {
				lastErr = errors.New("deepseek returned no choices")
			} else {
				content := decoded.Choices[0].Message.Content
				c.logger.Printf("deepseek response attempt=%d content=%q", attempt, content)
				return content, nil
			}
		}

		if attempt < maxAttempts {
			sleep := time.Second * time.Duration(attempt)
			select {
			case <-time.After(sleep):
			case <-ctx.Done():
				return "", ctx.Err()
			}
		}
	}

	return "", lastErr
}

func (c *DeepSeekClient) WithLogger(logger Logger) *DeepSeekClient {
	if logger == nil {
		return c
	}
	c.logger = logger
	return c
}
